controller/internal/supervisor/iptablesctrl/acls_windows.go

```diff
@@ -0,0 +1,218 @@
+	}
+
+	for _, rule := range rules {
+		zap.L().Error("Creating rules", zap.Strings("rule", rule))
+		if err := i.impl.NewChain(rule[1], rule[3]); err != nil {
+			return fmt.Errorf("unable to add chain %s of context %s %s", rule[3], rule[1], err)
+		}
+	}
+
+	return nil
+}
+
+// deletePUChains removes all the container specific chains and basic rules
+func (i *iptables) deletePUChains(cfg *ACLInfo, containerInfo *policy.PUInfo) error {
+
+	// For Windows, instead of clearing and deleting the app-chain and the net-chain, we need to
+	// delete individual rules. This is because Windows uses non-PU-specific chains.
+
+	// delete ACLs rules for Windows
+	appACLIPset := i.getACLIPSets(containerInfo.Policy.ApplicationACLs())
+	netACLIPset := i.getACLIPSets(containerInfo.Policy.NetworkACLs())
+
+	if err := i.deleteExternalACLs(cfg, cfg.AppChain, cfg.NetChain, appACLIPset, true); err != nil {
+		zap.L().Warn("Failed to delete ACL rules for app chain", zap.Error(err))
+	}
+
+	if err := i.deleteExternalACLs(cfg, cfg.NetChain, cfg.AppChain, netACLIPset, false); err != nil {
+		zap.L().Warn("Failed to delete ACL rules for net chain", zap.Error(err))
+	}
+
+	// delete deny-all rules for Windows
+	isHostPU := extractors.IsHostPU(containerInfo.Runtime, i.mode)
+	return i.processRulesFromList(i.trapRules(cfg, isHostPU, [][]string{}, [][]string{}), "Delete")
+}
+
+func (i *iptables) deleteExternalACLs(cfg *ACLInfo, chain string, reverseChain string, rules []aclIPset, isAppAcls bool) error {
+
+	_, rules = extractProtocolAnyRules(rules)
+
+	rulesBucket := i.sortACLsInBuckets(cfg, chain, reverseChain, rules, isAppAcls)
+
+	aclRules, err := extractACLsFromTemplate(rulesBucket)
+	if err != nil {
+		return fmt.Errorf("unable to extract rules from template: %s", err)
+	}
+
+	aclRules = transformACLRules(aclRules, cfg, rulesBucket, isAppAcls)
+
+	if err := i.processRulesFromList(aclRules, "Delete"); err != nil {
+		return fmt.Errorf("unable to delete rules - mode :%s %v", err, isAppAcls)
+	}
+
+	return nil
+}
+
+// removeGlobalHooksPre for Windows does nothing
+func (i *iptables) removeGlobalHooksPre() {
+}
+
+// try to merge two acl rules (one log and one accept/drop) into one for Windows
+func makeTerminatingRuleFromPair(aclRule1, aclRule2 []string) *winipt.WindowsRuleSpec {
+
+	if aclRule1 == nil || aclRule2 == nil {
+		return nil
+	}
+	winRuleSpec1, err := winipt.ParseRuleSpec(aclRule1[2:]...)
+	if err != nil {
+		return nil
+	}
+	winRuleSpec2, err := winipt.ParseRuleSpec(aclRule2[2:]...)
+	if err != nil {
+		return nil
+	}
+
+	// save action/log properties
+	action := winRuleSpec1.Action
+	logPrefix := winRuleSpec2.LogPrefix
+	groupId := winRuleSpec2.GroupId
+	if action == 0 {
+		action = winRuleSpec2.Action
+		if action == 0 {
+			return nil
+		}
+		logPrefix = winRuleSpec1.LogPrefix
+		groupId = winRuleSpec1.GroupId
+		if !winRuleSpec1.Log {
+			return nil
+		}
+	} else if !winRuleSpec2.Log {
+		return nil
+	}
+
```

Comment:
I would avoid the if-then-else 

if !winRuleSpec2.Log {
  return nil 
}

Any place we have if then , else if .. should be optimized to a simpler structure or its impossible to follow.

Reply:
I have modified this code to try to make more sense. You still have to check a fair number of conditions, but I hope it makes more sense now. (I'm trying to check that one rule has an Action and the other has Log attributes, and if so then save the appropriate related attributes.)